packages/zoe/test/unitTests/contracts/test-coveredCall.js

```diff
@@ -1070,3 +1070,44 @@ test('zoe - coveredCall non-fungible', async t => {
   t.deepEqual(bobCcPurse.getCurrentAmount().value, ['GrowlTiger']);
   t.deepEqual(bobRpgPurse.getCurrentAmount().value, []);
 });
+
+test('zoe - coveredCall - bad proposal shape', async t => {
+  const { moolaKit, simoleanKit, moola, zoe, vatAdminState } = setup();
+
+  // Bundle and install the contract.
+  const bundle = await bundleSource(coveredCallRoot);
+  vatAdminState.installBundle('b1-coveredcall', bundle);
+  const coveredCallInstallation =
+    await E(zoe).installBundleID('b1-coveredcall');
+
+  // Start an instance.
+  const issuerKeywordRecord = harden({
+    UnderlyingAsset: moolaKit.issuer,
+    StrikePrice: simoleanKit.issuer,
+  });
+  const { creatorInvitation } = await E(zoe).startInstance(
+    coveredCallInstallation,
+    issuerKeywordRecord,
+  );
+
+  // Make an unacceptable proposal.
+  const badProposal = harden({
+    give: { UnderlyingAsset: moola(3n) },
+    exit: { waived: null },
+  });
+  const payments = harden({
+    UnderlyingAsset: moolaKit.mint.mintPayment(moola(3n)),
+  });
+  const badSeat = await E(zoe).offer(creatorInvitation, badProposal, payments);
+  await t.throwsAsync(
+    () => E(badSeat).getOfferResult(),
+    { message: /Must have missing properties \["afterDeadline"\]/ },
```

Comment:
The history is fascinating. Each of the checks has been refactored and cleaned up a few times. It looks like #3184 was the one that added the version with the explicit error message, but didn't drop the `mustMatch` branch. I think they verify the same thing, so I prefer the version with the more specific error.  But switching to that would remove your opportunity to test the Pattern.

I think it would be better (for Zoe) to have an explicit test of an `offerPattern`, and not just of a `mustMatch`.  I can do that in a day with a purpose-made trivial contract. I'll make an issue and add it to the list of Zoe cleanups.

Are there no other tests of `mustMatch`?  If that's missing, it should go in endo or `@agoric/store` if there isn't one.

Reply:
There are tests of `mustMatch` in [`@endo/patterns`](https://github.com/endojs/endo/blob/master/packages/patterns/test/test-patterns.js); this PR is more about covering the behavior associated with submitting an unacceptable proposal:
* `E(zoe).offer(invitation, badProposal, payments)` still creates a seat.
* `getOfferResult()` on that seat throws an error complaining about the proposal.
* The seat's payouts match its payments.

----

P.S. I've been fighting with the linter on this PR; it turns out that [`getPayout(keyword)` is typed](https://github.com/Agoric/agoric-sdk/blob/51399d19e8b32f794e2e0b6af990cc439d60dfaa/packages/zoe/src/zoeService/types.js#L200) to return a promise for a Payment, but the [implementation](https://github.com/Agoric/agoric-sdk/blob/51399d19e8b32f794e2e0b6af990cc439d60dfaa/packages/zoe/src/zoeService/zoeSeat.js#L275-L286) can return a promise for `undefined`—which it does in this case, and also more concerningly in cases where the argument is not a contract-defined keyword or isn't even a keyword at all (e.g., `await E(seat).getPayout('notAKeyword')`). But I bet a lot of code is not prepared to handle an `undefined` result... does this warrant an issue as well?

Reply:
> this PR is more about covering the behavior associated with submitting an unacceptable proposal

If the `Pattern` were in the invitation's `proposalShape`, then it wouldn't need to create the seat in order to return payments, since Zoe would just not escrow them, I think.  If the Pattern isn't checked until it's into the handler, then the payments have already been escrowed, and Zoe has to return a seat in order to return the payment.

This seems like an issue for any particular contract, and not a fact about Zoe's support for Patterns. I don't think this test is necessary, though it reinforces the need for the one in #8606.

> getPayout(keyword) is typed to return a promise for a Payment, but the implementation can return a promise for undefined ... But I bet a lot of code is not prepared to handle an undefined result... does this warrant an issue as well?

Maybe that's an argument that `getPayout()` should throw rather than returning a promise for undefined. The type system is certainly not smart enough to understand that if your argument is right, you shouldn't have to check for undefined. In any case, *yes*, if the implementation doesn't match the types, we should add an issue.

Reply:
#8610